### PR TITLE
Add a basic glossary to the docs

### PR DESCRIPTION
**Pull request by smite**
**Description of the Change:**

Added a basic glossary page to the docs.
I wasn't able to provide a great definition for some of the terms, which IMO highlights the need for a glossary :)

**Benefits:**

Makes it easier to keep the terminology clear.

### CODE REVIEW

File: doc/glossary.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+   operation
+   quantum operation
+      Something that acts on specific subsystems of the :term:`register`, changing its state and
+      possibly returning classical information, e.g. a state preparation, :term:`gate`,
+      :term:`measurement` etc.
+      See: :class:`.Operation`
+
+   register
+   quantum register
```
Comment (josh146) [inline doc/glossary.rst]: Should we add 'subsystems' and 'modes' here as well --- we tend to use all four throughout the documentation in different places

File: doc/glossary.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+   equivalent circuit
+   equivalent program
+      Two quantum :term:`circuits <circuit>` are equivalent iff they produce the same output
+      probability distributions.
+
+   engine
+   frontend
+      A class for executing quantum :term:`programs <program>` on a specific :term:`backend`.
+      See: :class:`.BaseEngine`
```
[the remark this answers is not in the document]
Comment (smite) [reply doc/glossary.rst]: Yeah, I guess so. Let's remove frontend from the glossary for now.

File: doc/glossary.rst (hunk 1 above)
Comment (smite) [reply doc/glossary.rst]: right

File: doc/glossary.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      See: :class:`.Program`
+
+   compilation
+      The process of converting a quantum :term:`program` into an :term:`equivalent program` that
+      can be run on a specific :term:`device`. The compilation can fail if the device is incapable
+      of running the program.
+
+   device
+      Something that can run a well-defined subclass of quantum :term:`programs <program>`.
+      Each :term:`backend` corresponds to one or more devices.
```
Comment (co9olguy) [inline doc/glossary.rst]:
```suggestion
      The same device can potentially be evaluated on more than one :term:`backend`, and each :term:`backend` can implement one or more devices.
```

File: doc/glossary.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      Typically contains :term:`measurements <measurement>` that produce a stochastic classical output.
+      Represents a quantum computation.
+      See: :class:`.Program`
+
+   compilation
+      The process of converting a quantum :term:`program` into an :term:`equivalent program` that
+      can be run on a specific :term:`device`. The compilation can fail if the device is incapable
+      of running the program.
+
+   device
```
Comment (co9olguy) [inline doc/glossary.rst]: @smite @josh146 the name "device" is definitely the most unclear concept. Suggest choosing better name

Possibilities:
1. circuit family
1. circuit type
1. circuit class
1. program spec
1. subprogram
1. program family
1. program type
1. program class

File: doc/glossary.rst [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      Each :term:`backend` corresponds to one or more devices.
+      See: :class:`.DeviceSpecs`
+
+   equivalent circuit
+   equivalent program
+      Two quantum :term:`circuits <circuit>` are equivalent iff they produce the same output
+      probability distributions.
+
+   engine
+   frontend
```
Comment (co9olguy) [inline doc/glossary.rst]:
```suggestion
```

File: doc/glossary.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply doc/glossary.rst]: no need to remove, let's just add another glossary term, with content similar to what @josh146 wrote above

File: doc/glossary.rst (hunk 1 above)
Comment (co9olguy) [inline doc/glossary.rst]:
```suggestion
   quantum register
   system
   mode
   qumode
```

File: doc/glossary.rst [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      A quantum :term:`operation` that changes the state of the :term:`register`.
+      Analogous to a logic gate in a Boolean circuit.
+
+   measurement
+      A quantum :term:`operation` that returns classical information.
+
+   operation
+   quantum operation
+      Something that acts on specific subsystems of the :term:`register`, changing its state and
+      possibly returning classical information, e.g. a state preparation, :term:`gate`,
```
Comment (co9olguy) [inline doc/glossary.rst]:
```suggestion
      possibly returning classical information, e.g., a state preparation, :term:`gate`,
```

File: doc/glossary.rst [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      :term:`measurement` etc.
```
Comment (co9olguy) [inline doc/glossary.rst]:
```suggestion
      :term:`measurement`, etc.
```

File: doc/glossary.rst [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      See: :class:`.Operation`
+
+   register
+   quantum register
+      A quantum system (or a simulation of one) storing the state of the quantum computation.
+      Typically consists of several subsystems (which in CV quantum computing are called modes).
```
Comment (co9olguy) [inline doc/glossary.rst]: could add `state preparation`, but not urgent

File: doc/glossary.rst (hunk 4 above)
Comment (josh146) [reply doc/glossary.rst]: Of the above, I think I like 'circuit class' or 'circuit type' the most, but it still doesn't sound quite right...

Comment (smite) [reply doc/glossary.rst]: I guess "circuit template" is already taken, and means a parametrized circuit?

**Comment by smite** [comment]
Man, I feel bad when each time I commit a small documentation change Travis runs all the tests again :|

**Comment by josh146** [comment]
@smite, you can add `[ci skip]` to the commit message, and Travis will be skipped 🙂

File: doc/glossary.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply doc/glossary.rst]: done

**Pull request merged**